### PR TITLE
[Fix] Fix binary segmentation when `num_classes==1`

### PR DESCRIPTION
**Pull request by xiexinch**
## Motivation

Fix issue #1774

## Modification

1. Add `threshold` and `out_channels` parameters to `decode_head`
2. Fix inference logits while `num_classes==1`

## BC-breaking (Optional)

None

### CODE REVIEW

File: mmseg/models/segmentors/encoder_decoder.py [hunk 1]
```diff
@@ -245,7 +245,10 @@ def inference(self, img, img_meta, rescale):
             seg_logit = self.slide_inference(img, img_meta, rescale)
         else:
             seg_logit = self.whole_inference(img, img_meta, rescale)
-        output = F.softmax(seg_logit, dim=1)
+        if self.num_classes == 1:
```
Comment (MeowZheng) [inline mmseg/models/segmentors/encoder_decoder.py:248]:
```suggestion
        if self.out_channels == 1:
```

File: mmseg/models/segmentors/encoder_decoder.py [hunk 2]
```diff
@@ -260,7 +263,11 @@ def inference(self, img, img_meta, rescale):
     def simple_test(self, img, img_meta, rescale=True):
         """Simple test with single image."""
         seg_logit = self.inference(img, img_meta, rescale)
-        seg_pred = seg_logit.argmax(dim=1)
+        if self.num_classes == 1:
```
Comment (MeowZheng) [inline mmseg/models/segmentors/encoder_decoder.py:266]: same problems as above

File: mmseg/models/segmentors/encoder_decoder.py [hunk 3]
```diff
@@ -283,7 +290,11 @@ def aug_test(self, imgs, img_metas, rescale=True):
             cur_seg_logit = self.inference(imgs[i], img_metas[i], rescale)
             seg_logit += cur_seg_logit
         seg_logit /= len(imgs)
-        seg_pred = seg_logit.argmax(dim=1)
+        if self.num_classes == 1:
```
Comment (MeowZheng) [inline mmseg/models/segmentors/encoder_decoder.py:293]: same problems as above

File: mmseg/models/decode_heads/decode_head.py [hunk 4]
```diff
@@ -84,6 +91,12 @@ def __init__(self,
         self.ignore_index = ignore_index
         self.align_corners = align_corners
 
+        if self.num_classes == 2 and self.out_channels is None:
+            warnings.warn('set out_channels to 1 to reduce the number of model\
+                parameters')
+        elif self.num_classes == 2 and self.out_channels == 1:
+            self.num_classes = 1
```
Comment (MeowZheng) [inline mmseg/models/decode_heads/decode_head.py:98]: if self. out_channels is None:
    self.out_channels = self.num_classes
    if self.num_classes == 2:
        warning.warn('For binary segmentation, we suggest to use `out_channels = 1` to define the output channels of segmentor', and use `threshold` to  convert seg_logist into a prediction applying a threshold

File: mmseg/models/decode_heads/decode_head.py [hunk 5]
```diff
@@ -18,6 +19,8 @@ class BaseDecodeHead(BaseModule, metaclass=ABCMeta):
         in_channels (int|Sequence[int]): Input channels.
         channels (int): Channels after modules, before conv_seg.
         num_classes (int): Number of classes.
+        threshold (float): Threshold for binary segmentation in the case of
+            `num_classes==1`. Default: 0.3.
```
Comment (MeowZheng) [inline mmseg/models/decode_heads/decode_head.py:23]: threshold default to None.

File: mmseg/models/decode_heads/decode_head.py [hunk 6]
```diff
@@ -75,6 +80,8 @@ def __init__(self,
         self._init_inputs(in_channels, in_index, input_transform)
         self.channels = channels
         self.num_classes = num_classes
+        self.out_channels = out_channels
+        self.threshold = threshold
```
Comment (MeowZheng) [inline mmseg/models/decode_heads/decode_head.py:84]:
```suggestion
```

File: mmseg/models/decode_heads/decode_head.py [hunk 7]
```diff
@@ -56,6 +59,8 @@ def __init__(self,
                  channels,
                  *,
                  num_classes,
+                 out_channels=None,
+                 threshold=0.3,
```
Comment (MeowZheng) [inline mmseg/models/decode_heads/decode_head.py:63]: threshold = None.

File: mmseg/models/decode_heads/decode_head.py [hunk 8]
```diff
@@ -84,6 +89,29 @@ def __init__(self,
...
+                              '`out_channels = 1` to define the output'
+                              'channels of segmentor, and use `threshold`'
+                              'to convert seg_logist into a prediction'
+                              'applying a threshold')
+            out_channels = num_classes
+
+        if out_channels != num_classes and out_channels != 1:
+            raise ValueError(
+                'out_channels should be equal to num_classes,'
+                'except out_channels == 1 and num_classes == 2, but got'
```
Comment (MeowZheng) [inline mmseg/models/decode_heads/decode_head.py:104]:
```suggestion
                'except binary segmentation set out_channels == 1 and num_classes == 2, but got'
```

**Review by MeowZheng — APPROVED** [review]
